src/module_3/assign_3.md

```diff
@@ -0,0 +1,980 @@
+      <td>30.0</td>
+      <td>30.0</td>
+      <td>24.27618</td>
+    </tr>
+    <tr>
+      <th>4</th>
+      <td>0</td>
+      <td>33826472919172</td>
+      <td>2808429314180</td>
+      <td>3537167515780</td>
+      <td>128098</td>
+      <td>1.601942e+09</td>
+      <td>3</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>...</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>2.0</td>
+      <td>33.0</td>
+      <td>42.0</td>
+      <td>31.134053</td>
+      <td>30.0</td>
+      <td>30.0</td>
+      <td>24.27618</td>
+    </tr>
+  </tbody>
+</table>
+<p>5 rows × 26 columns</p>
+</div>
+
+
+
+
+```python
+#Filtering for transactions with more than 5 items sold
+df_orders = df.groupby(by='order_id')['outcome'].sum()
+print(len(df_orders))
+transac_5_items = df_orders[df_orders>5].index.to_list()
+```
+
+    3446
+
+
+
+```python
+#Splitting for test/train/validation, keeping the transactions in the same dataset to avoid data leakeage
+
+random.shuffle(transac_5_items)
+
+train_data_orders =  transac_5_items[:int(len(transac_5_items)*0.7)]
+val_data_orders = transac_5_items[int(len(transac_5_items)*0.7):int(len(transac_5_items)*0.9)]
+test_data_orders = transac_5_items[int(len(transac_5_items)*0.9):]
+```
+
+### Main Models
+
+
+```python
+train_data_x = df[df['order_id'].isin(train_data_orders)].iloc[:,4:]
+train_data_y = df[df['order_id'].isin(train_data_orders)].iloc[:,0]
+val_data_x = df[df['order_id'].isin(val_data_orders)].iloc[:,4:]
+val_data_y = df[df['order_id'].isin(val_data_orders)].iloc[:,0]
+test_data_x = df[df['order_id'].isin(test_data_orders)].iloc[:,4:]
+test_data_y = df[df['order_id'].isin(test_data_orders)].iloc[:,0]
+```
+
+
+```python
+def draw_roc_prc(real_y, y_prob, title):
+    '''
+    Plots a ROC and a Precision and Recall curve.
+    '''
+    fig, ax = plt.subplots(1, 2, figsize=(10, 4))
+    fpr, tpr, thresholds = metrics.roc_curve(real_y, y_prob)
+    roc_auc = metrics.auc(fpr, tpr)
+    
+    fig.suptitle(f'ROC and Precision-Recall Curves: {title}')
+
+    ax[0].set_title(f'ROC curve (AUC = {roc_auc:.2f})')
+    ax[0].plot(fpr, tpr, color='darkorange', lw=2)
+    ax[0].plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--')
+    ax[0].set_xlim([0.0, 1.0])
+    ax[0].set_ylim([0.0, 1.05])
+    ax[0].set_xlabel('False Positive Rate')
+    ax[0].set_ylabel('True Positive Rate')
+
+    precision, recall, thresholds = metrics.precision_recall_curve(real_y, y_prob)
+
+    average_precision = metrics.average_precision_score(real_y, y_prob)
+
+    ax[1].plot(recall, precision, marker='.')
+    ax[1].set_xlabel('Recall')
+    ax[1].set_ylabel('Precision')
+    ax[1].set_title('Precision-Recall Curve (AUC-PR = {:.2f})'.format(average_precision))
+    plt.tight_layout()
+    plt.show()
+
+
+```
+
+
+```python
+#Base model with no penalties
+log_model_base = LogisticRegression(penalty= None, n_jobs=-1, class_weight='balanced').fit(train_data_x,train_data_y)
```

Comment:
As we discuss in class, this balancing is "stacking" many points on each positive sample instance (to give them way more weight to correct for the balance). A potential problem is that we are over representing the specific points that happen to be sampled (from the underlying phenomenon) in my dataset.

Reply:
I removed this parameter. I get that if our input dataset is going to look as unbalanced as the training one, it doesn't make sense to balance the classes in training.